// @botpress/cli@0.6.16/dist/api/bot-body.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bot_body_exports = {};
__export(bot_body_exports, {
  prepareCreateBotBody: () => prepareCreateBotBody,
  prepareUpdateBotBody: () => prepareUpdateBotBody
});
module.exports = __toCommonJS(bot_body_exports);
var utils = __toESM(require("../utils"));
const prepareCreateBotBody = (bot) => ({
  ...bot.props,
  configuration: bot.props.configuration ? {
    ...bot.props.configuration,
    schema: utils.schema.mapZodToJsonSchema(bot.props.configuration)
  } : void 0,
  events: bot.props.events ? utils.records.mapValues(bot.props.events, (event) => ({
    ...event,
    schema: utils.schema.mapZodToJsonSchema(event)
  })) : void 0,
  states: bot.props.states ? utils.records.mapValues(bot.props.states, (state) => ({
    ...state,
    schema: utils.schema.mapZodToJsonSchema(state)
  })) : void 0
});
const prepareUpdateBotBody = (localBot, remoteBot) => ({
  ...localBot,
  states: utils.records.setNullOnMissingValues(localBot.states, remoteBot.states),
  recurringEvents: utils.records.setNullOnMissingValues(localBot.recurringEvents, remoteBot.recurringEvents),
  events: utils.records.setNullOnMissingValues(localBot.events, remoteBot.events),
  user: {
    ...localBot.user,
    tags: utils.records.setNullOnMissingValues(localBot.user?.tags, remoteBot.user?.tags)
  },
  conversation: {
    ...localBot.conversation,
    tags: utils.records.setNullOnMissingValues(localBot.conversation?.tags, remoteBot.conversation?.tags)
  },
  message: {
    ...localBot.message,
    tags: utils.records.setNullOnMissingValues(localBot.message?.tags, remoteBot.message?.tags)
  },
  integrations: utils.records.setNullOnMissingValues(localBot.integrations, remoteBot.integrations)
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  prepareCreateBotBody,
  prepareUpdateBotBody
});
