// @botpress/cli@0.6.16/dist/api/client.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var client_exports = {};
__export(client_exports, {
  ApiClient: () => ApiClient
});
module.exports = __toCommonJS(client_exports);
var client = __toESM(require("@botpress/client"));
var import_lodash = __toESM(require("lodash"));
var import_integration_ref = require("../integration-ref");
var paging = __toESM(require("./paging"));
class ApiClient {
  constructor(props, _logger) {
    this._logger = _logger;
    const { apiUrl, token, workspaceId } = props;
    this.client = new client.Client({ apiUrl, token, workspaceId });
    this.url = apiUrl;
    this.token = token;
    this.workspaceId = workspaceId;
  }
  client;
  url;
  token;
  workspaceId;
  static newClient = (props, logger) => new ApiClient(props, logger);
  async findIntegration(ref) {
    const formatted = (0, import_integration_ref.formatIntegrationRef)(ref);
    const privateIntegration = await this.findPrivateIntegration(ref);
    if (privateIntegration) {
      this._logger.debug(`Found integration "${formatted}" in workspace`);
      return privateIntegration;
    }
    const publicIntegration = await this.findPublicIntegration(ref);
    if (publicIntegration) {
      this._logger.debug(`Found integration "${formatted}" in hub`);
      return publicIntegration;
    }
    return;
  }
  async findPrivateIntegration(ref) {
    const { workspaceId } = this;
    if (ref.type === "id") {
      return this.validateStatus(
        () => this.client.getIntegration(ref).then((r) => ({ ...r.integration, workspaceId })),
        404
      );
    }
    return this.validateStatus(
      () => this.client.getIntegrationByName(ref).then((r) => ({ ...r.integration, workspaceId })),
      404
    );
  }
  async findPublicIntegration(ref) {
    if (ref.type === "id") {
      return this.validateStatus(() => this.client.getPublicIntegrationById(ref).then((r) => r.integration), 404);
    }
    return this.validateStatus(() => this.client.getPublicIntegration(ref).then((r) => r.integration), 404);
  }
  async testLogin() {
    await this.client.listBots({});
  }
  listAllPages = paging.listAllPages;
  async validateStatus(fn, allowedStatuses) {
    try {
      const v = await fn();
      return v;
    } catch (err) {
      const allowedStatusesArray = import_lodash.default.isArray(allowedStatuses) ? allowedStatuses : [allowedStatuses];
      const isAllowed = client.isApiError(err) && err.code && allowedStatusesArray.includes(err.code);
      if (isAllowed) {
        return;
      }
      throw err;
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApiClient
});
