// @botpress/cli@0.6.16/dist/consts.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var consts_exports = {};
__export(consts_exports, {
  cliRootDir: () => cliRootDir,
  defaultBotpressApiUrl: () => defaultBotpressApiUrl,
  defaultBotpressAppUrl: () => defaultBotpressAppUrl,
  defaultBotpressHome: () => defaultBotpressHome,
  defaultEntrypoint: () => defaultEntrypoint,
  defaultOutputFolder: () => defaultOutputFolder,
  defaultTunnelUrl: () => defaultTunnelUrl,
  echoBotDirName: () => echoBotDirName,
  emptyIntegrationDirName: () => emptyIntegrationDirName,
  fromCliRootDir: () => fromCliRootDir,
  fromHomeDir: () => fromHomeDir,
  fromOutDir: () => fromOutDir,
  fromWorkDir: () => fromWorkDir
});
module.exports = __toCommonJS(consts_exports);
var import_os = __toESM(require("os"));
var import_path = __toESM(require("path"));
var import_root = require("./root");
const defaultBotpressHome = import_path.default.join(import_os.default.homedir(), ".botpress");
const defaultOutputFolder = ".botpress";
const defaultEntrypoint = import_path.default.join("src", "index.ts");
const defaultBotpressApiUrl = "https://api.botpress.cloud";
const defaultBotpressAppUrl = "https://app.botpress.cloud";
const defaultTunnelUrl = "https://tunnel.botpress.cloud";
const cliRootDir = import_root.CLI_ROOT_DIR;
const echoBotDirName = "echo-bot";
const emptyIntegrationDirName = "empty-integration";
const fromCliRootDir = {
  echoBotTemplate: import_path.default.join("templates", echoBotDirName),
  emptyIntegrationTemplate: import_path.default.join("templates", emptyIntegrationDirName)
};
const fromHomeDir = {
  globalCacheFile: "global.cache.json"
};
const fromWorkDir = {
  definition: "integration.definition.ts"
};
const fromOutDir = {
  distDir: "dist",
  outFile: import_path.default.join("dist", "index.js"),
  installDir: "installations",
  implementationDir: "implementation",
  secretsDir: "secrets",
  projectCacheFile: "project.cache.json"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  cliRootDir,
  defaultBotpressApiUrl,
  defaultBotpressAppUrl,
  defaultBotpressHome,
  defaultEntrypoint,
  defaultOutputFolder,
  defaultTunnelUrl,
  echoBotDirName,
  emptyIntegrationDirName,
  fromCliRootDir,
  fromHomeDir,
  fromOutDir,
  fromWorkDir
});
