// @botpress/cli@0.6.16/templates/empty-integration/.botpress/implementation/channels/channel/messages/index.ts
/* tslint:disable */
// This file is generated
// Do not edit this file
import * as text from "./text";
export * as text from "./text";
import * as markdown from "./markdown";
export * as markdown from "./markdown";
import * as image from "./image";
export * as image from "./image";
import * as audio from "./audio";
export * as audio from "./audio";
import * as video from "./video";
export * as video from "./video";
import * as file from "./file";
export * as file from "./file";
import * as location from "./location";
export * as location from "./location";
import * as carousel from "./carousel";
export * as carousel from "./carousel";
import * as card from "./card";
export * as card from "./card";
import * as dropdown from "./dropdown";
export * as dropdown from "./dropdown";
import * as choice from "./choice";
export * as choice from "./choice";

export type Messages = {
  text: text.Text;
  markdown: markdown.Markdown;
  image: image.Image;
  audio: audio.Audio;
  video: video.Video;
  file: file.File;
  location: location.Location;
  carousel: carousel.Carousel;
  card: card.Card;
  dropdown: dropdown.Dropdown;
  choice: choice.Choice;
}
