// @botpress/cli@0.6.16/dist/worker/config.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var config_exports = {};
__export(config_exports, {
  CONFIG_ENV_KEY: () => CONFIG_ENV_KEY,
  configSchema: () => configSchema
});
module.exports = __toCommonJS(config_exports);
var import_zod = __toESM(require("zod"));
const CONFIG_ENV_KEY = "CHILD_PROCESS_CONFIGURATION";
const configSchema = import_zod.default.union([
  import_zod.default.object({
    type: import_zod.default.literal("code"),
    code: import_zod.default.string(),
    env: import_zod.default.record(import_zod.default.string(), import_zod.default.string()).optional()
  }),
  import_zod.default.object({
    type: import_zod.default.literal("file"),
    file: import_zod.default.string(),
    env: import_zod.default.record(import_zod.default.string(), import_zod.default.string()).optional()
  })
]);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CONFIG_ENV_KEY,
  configSchema
});
