// @botpress/cli@0.6.16/dist/code-generation/module.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var module_exports = {};
__export(module_exports, {
  Module: () => Module,
  ReExportTypeModule: () => ReExportTypeModule
});
module.exports = __toCommonJS(module_exports);
var import_path = require("path");
var utils = __toESM(require("../utils"));
var import_const = require("./const");
class Module {
  constructor(_def) {
    this._def = _def;
  }
  _localDependencies = [];
  get path() {
    return this._def.path;
  }
  get content() {
    return this._def.content;
  }
  get name() {
    const basename = import_path.posix.basename(this.path);
    if (basename === import_const.INDEX_FILE) {
      const dirname = import_path.posix.basename(import_path.posix.dirname(this.path));
      return utils.casing.to.camelCase(dirname);
    }
    const withoutExtension = utils.path.rmExtension(basename);
    return utils.casing.to.camelCase(withoutExtension);
  }
  get exports() {
    return this._def.exportName;
  }
  get deps() {
    return [...this._localDependencies];
  }
  pushDep(...dependencies) {
    this._localDependencies.push(...dependencies);
    return this;
  }
  unshift(...basePath) {
    this._def = {
      ...this._def,
      path: import_path.posix.join(...basePath, this._def.path)
    };
    this._localDependencies = this._localDependencies.map((d) => d.unshift(...basePath));
    return this;
  }
  flatten() {
    return [this, ...this._localDependencies.flatMap((d) => d.flatten())];
  }
  import(base) {
    let relativePath = import_path.posix.relative(import_path.posix.dirname(base.path), this.path);
    relativePath = import_path.posix.join(".", relativePath);
    return utils.path.rmExtension(relativePath);
  }
}
class ReExportTypeModule extends Module {
  constructor(def) {
    super({
      ...def,
      path: import_const.INDEX_FILE,
      content: ""
    });
  }
  get content() {
    let content = import_const.GENERATED_HEADER;
    const dependencies = this.deps;
    for (const m of dependencies) {
      const { name } = m;
      const importFrom = m.import(this);
      content += `import * as ${name} from "./${importFrom}";
`;
      content += `export * as ${name} from "./${importFrom}";
`;
    }
    content += "\n";
    content += `export type ${this.exports} = {
`;
    for (const { name, exports } of this.deps) {
      content += `  ${name}: ${name}.${exports};
`;
    }
    content += "}";
    content += "\n";
    return content;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Module,
  ReExportTypeModule
});
