// @botpress/cli@0.6.16/dist/utils/esbuild-utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var esbuild_utils_exports = {};
__export(esbuild_utils_exports, {
  buildCode: () => buildCode,
  buildEntrypoint: () => buildEntrypoint
});
module.exports = __toCommonJS(esbuild_utils_exports);
var import_esbuild = require("esbuild");
const keepNames = true;
function buildCode({
  cwd,
  minify = true,
  bundle = true,
  sourcemap = false,
  logLevel = "silent",
  outfile,
  code,
  write
}) {
  return (0, import_esbuild.build)({
    stdin: {
      contents: code,
      resolveDir: cwd,
      loader: "ts"
    },
    logOverride: {
      "equals-negative-zero": "silent"
    },
    platform: "node",
    target: "es2020",
    sourcemap,
    minify,
    bundle,
    outfile,
    absWorkingDir: cwd,
    logLevel,
    keepNames,
    write
  });
}
function buildEntrypoint({
  cwd,
  minify = true,
  bundle = true,
  sourcemap = false,
  logLevel = "silent",
  outfile,
  entrypoint,
  write
}) {
  return (0, import_esbuild.build)({
    entryPoints: [entrypoint],
    logOverride: {
      "equals-negative-zero": "silent"
    },
    platform: "node",
    target: "es2020",
    sourcemap,
    minify,
    bundle,
    outfile,
    absWorkingDir: cwd,
    logLevel,
    keepNames,
    write
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  buildCode,
  buildEntrypoint
});
