// @botpress/cli@0.6.16/dist/utils/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  cache: () => cache,
  casing: () => casing,
  emitter: () => emitter,
  esbuild: () => esbuild,
  filewatcher: () => filewatcher,
  guards: () => guards,
  id: () => id,
  path: () => path,
  pkgJson: () => pkgJson,
  prompt: () => prompt,
  records: () => records,
  require: () => require2,
  schema: () => schema,
  semver: () => semver,
  string: () => string,
  tunnel: () => tunnel,
  types: () => types,
  url: () => url
});
module.exports = __toCommonJS(utils_exports);
var cache = __toESM(require("./cache-utils"));
var casing = __toESM(require("./case-utils"));
var emitter = __toESM(require("./event-emitter"));
var esbuild = __toESM(require("./esbuild-utils"));
var filewatcher = __toESM(require("./file-watcher"));
var guards = __toESM(require("./guard-utils"));
var id = __toESM(require("./id-utils"));
var path = __toESM(require("./path-utils"));
var pkgJson = __toESM(require("./pkgjson-utils"));
var prompt = __toESM(require("./prompt-utils"));
var records = __toESM(require("./record-utils"));
var require2 = __toESM(require("./require-utils"));
var schema = __toESM(require("./schema-utils"));
var semver = __toESM(require("./semver-utils"));
var string = __toESM(require("./string-utils"));
var tunnel = __toESM(require("./tunnel-utils"));
var types = __toESM(require("./type-utils"));
var url = __toESM(require("./url-utils"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  cache,
  casing,
  emitter,
  esbuild,
  filewatcher,
  guards,
  id,
  path,
  pkgJson,
  prompt,
  records,
  require,
  schema,
  semver,
  string,
  tunnel,
  types,
  url
});
